Lib/idlelib/idle_test/test_sidebar.py

```diff
@@ -370,5 +391,326 @@ def assert_colors_are_equal(colors):
+            cls.shell = None
+        cls.flist = None
+        cls.root.update_idletasks()
+        cls.root.destroy()
+        cls.root = None
+
+    @classmethod
+    def init_shell(cls):
+        cls.shell = cls.flist.open_shell()
+        cls.shell.pollinterval = 10
+        cls.root.update()
+        cls.n_preface_lines = get_lineno(cls.shell.text, 'end-1c') - 1
+
+    @classmethod
+    def reset_shell(cls):
+        cls.shell.per.bottom.delete(f'{cls.n_preface_lines+1}.0', 'end-1c')
+        cls.shell.shell_sidebar.update_sidebar()
+        cls.root.update()
+
+    def setUp(self):
+        if self.shell is None:
+            self.init_shell()
+        else:
+            self.reset_shell()
+
+    def get_sidebar_lines(self):
+        canvas = self.shell.shell_sidebar.canvas
+        texts = list(canvas.find(tk.ALL))
+        texts.sort(key=lambda text: canvas.bbox(text)[1])
+        return [canvas.itemcget(text, 'text') for text in texts]
+
+    def assert_sidebar_lines_end_with(self, expected_lines):
+        self.shell.shell_sidebar.update_sidebar()
+        self.assertEqual(
+            self.get_sidebar_lines()[-len(expected_lines):],
+            expected_lines,
+        )
+
+    def get_shell_line_y_coords(self):
+        text = self.shell.text
+        y_coords = []
+        index = text.index("@0,0")
+        while True:
+            lineinfo = text.dlineinfo(index)
+            if lineinfo is None:
+                break
+            y_coords.append(lineinfo[1])
+            index = text.index(f"{index} +1line")
+        return y_coords
+
+    def get_sidebar_line_y_coords(self):
+        canvas = self.shell.shell_sidebar.canvas
+        texts = list(canvas.find(tk.ALL))
+        texts.sort(key=lambda text: canvas.bbox(text)[1])
+        return [canvas.bbox(text)[1] for text in texts]
+
+    def assert_sidebar_lines_synced(self):
+        self.assertEqual(
+            self.get_sidebar_line_y_coords(),
+            self.get_shell_line_y_coords(),
+        )
+
+    def do_input(self, input):
+        shell = self.shell
+        text = shell.text
+        for line_index, line in enumerate(input.split('\n')):
+            if line_index > 0:
+                text.event_generate('<<newline-and-indent>>')
+            for char in line:
+                text.insert('insert', char)
+
+    def run_test_coroutine(self, coroutine):
+        root = self.root
+        # Exceptions raised by self.assert...() need to be raised outside of
+        # the after callback in order for the test harness to capture them.
+        exception = None
+        def after_callback():
+            nonlocal exception
+            try:
+                interval_multiplier = next(coroutine) or 1
+            except StopIteration:
+                root.quit()
+            except Exception as exc:
+                exception = exc
+                root.quit()
+            else:
+                root.after(100 * interval_multiplier, after_callback)
```

Comment:
Could we use `.after_idle` instead?

Reply:
Unfortunately, no. We need to wait for IDLE to send commands to its sub-process via the RPC mechanism, poll the RPC mechanism for a response, update stuff as per the response, and only then run the next piece of code.

Reply:
We *could* try running IDLE in these tests without a sub-process, then try `.after_idle()` or shorter timeouts.

Reply:
Hey, that worked! And reduce run time to under 1 second! :)